### DISCUSSION TITLE
Creating a SNN model on numerical data

### DISCUSSION THREAD

**Discussion by mahsa-ebrahimian** [BindsNET/bindsnet#451] [Q&A]
Hi All
I am totally new to Spiking neural network and do my research in this area. I am going to use bindsnet. I have been trying to find a sample code for a classification model and came up with below article:

[https://www.frontiersin.org/files/Articles/409297/fninf-12-00089-HTML/image_m/fninf-12-00089-g005.jpg](url)

Here is my code and I am facing errors on network run.

[https://github.com/mahsa-ebrahimian/Shilling_Attack_Detection/blob/main/Simple_Supervised.ipynb](url)

Could you please help me with the code or let me know if there is any example which has used a numerical dataset (not image) with SNN classification model.

**Comment by TheAwesomeAndy** [comment]
Hello,
Can you link the paper; it seems that the link provided is only a particular image from the paper

**Reply by mahsa-ebrahimian** [reply]
> Hello,
> Can you link the paper; it seems that the link provided is only a particular image from the paper

Sure, here is the Link to the article by @Hananel-Hazan 
[https://www.frontiersin.org/articles/10.3389/fninf.2018.00089/full](url)

**Reply by TheAwesomeAndy** [reply]
Thanks!  I'm working on a similar project involving coding a non-image dataset.  If I come across anything I'll post it here.  What coding sheme are you working with -- temporal / rate based?

**Reply by mahsa-ebrahimian** [reply]
> Thanks! I'm working on a similar project involving coding a non-image dataset. If I come across anything I'll post it here. What coding sheme are you working with -- temporal / rate based?

Sound Great! I am going to use temporal based, as I think its better based on my dataset.

**Comment by Hananel-Hazan** [comment]
The MNIST dateset in the above example has been normalize to value between 0 to 255. Then the encoder use the pixel value and convert it to spike train. 
Similar process you can do to any numerical value.

**Comment by hockman1** [comment]
Hi, the link above seems to be broken, COuld u reshare?

**Reply by mahsa-ebrahimian** [reply]
> Hi, the link above seems to be broken, COuld u reshare?

Could you please copy and paste it in URL. It will work.

https://www.frontiersin.org/articles/10.3389/fninf.2018.00089/full

**Comment by mahsa-ebrahimian** [comment]
I have fed my data into the SNN model in the supervised_mnist.py, underneath bindsnet examples. The problem is that it is giving me 0 precision and recall. As my dataset is imbalanced and I have two classes of 0 and 1 which is skewed toward 0s, what is your suggestion for improving the training model to achieve higher precision?
@Hananel-Hazan  @TheAwesomeAndy

**Reply by Hananel-Hazan** [reply]
Try to plot the activity of the network, check how many spikes the input produce.

**Reply by mahsa-ebrahimian** [reply]
> Try to plot the activity of the network, check how many spikes the input produce.

Thanks for your response.
I tried to plot the spikes diagram using bindsnet.analysis.plotting.plot_spikes, it seems there number of spikes are not many.
here is my code and you can see the diagram.

https://github.com/mahsa-ebrahimian/Shilling_Attack_Detection/blob/main/Supervised_on_MovieLens.ipynb

Is it because of my imbalanced data classes. What's your suggestion in this case for improving model's performance?

**Reply by Hananel-Hazan** [reply]
Nice advancement! 
Currently you are getting 93% accuracy, that is good.

Try adjust the normalization to reflect the data, also play with the nu value.  That may help to gain more accuracy

**Reply by mahsa-ebrahimian** [reply]
> Nice advancement!
> Currently you are getting 93% accuracy, that is good.
> 
> Try adjust the normalization to reflect the data, also play with the nu value. That may help to gain more accuracy

Thank you so much!
I added normalization added normalization and now I am getting better precision, but to find the best nu value ( learning rate), I need to plot training vs test loss function. But in SNN it seems we don't have predictions for each sample, is it right?
My second question is, if there is a way to change the activation function in the model, because my recall is almost constant during training.
@Hananel-Hazan

**Reply by Hananel-Hazan** [reply]
Good! 
I think you can modify the exist predictions function for use on one sample. 
What do you mean activation function?

**Comment by mahsa-ebrahimian** [comment]
Is there a way to add precision and recall of model to be considered during training process? Currently labels are updated based on accuracy only, It would be great to update them based on accuracy, precision and recall all together.

@Hananel-Hazan @hockman1  @amolk @TheAwesomeAndy

**Reply by Hananel-Hazan** [reply]
I believe it can be done, try to change the `evaluation.py` to fit your needs.
